Add endpoint to upload AI topic & increase max topic length

kathy-t opened the pull request:
**Description**
Related PRs: 
- https://github.com/dockstore/dockstore-support/pull/489
- https://github.com/dockstore/dockstore-ui2/pull/1950

This PR adds an extended TRS endpoint that uploads an AI topic for the entry identified by the TRS ID. If the manual and automatic topics are empty, then the topic selection is automatically set to AI. This new endpoint requires an admin or curator token. 

I also removed for end users to update the `topicAI` field using the `update` endpoint since AI topics are updated by the topic generator.

This PR also increases the maximum topic length from 150 to 250 characters because there were AI generated topics that exceeded 150 characters. I don't see any harm in doing this because the UI truncates the topics if they exceed 2 lines on the search page. I created a UI PR to update the maxLength allowed for the manual topic to 250 https://github.com/dockstore/dockstore-ui2/pull/1950.

**Review Instructions**
See instructions in https://github.com/dockstore/dockstore-support/pull/489

**Issue**
[SEAB-6007](https://ucsc-cgl.atlassian.net/browse/SEAB-6007)

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here and check the box when complete. 

- [x] Security and Privacy assessed

e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead. 


[SEAB-6007]: https://ucsc-cgl.atlassian.net/browse/SEAB-6007?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kathy-t commented:
Belated thought: should the endpoint automatically set the `TopicSelection` as AI if none of the other topics are filled in?

denis-yuen commented:
> Belated thought: should the endpoint automatically set the `TopicSelection` as AI if none of the other topics are filled in?

Makes sense, I guess they wouldn't be visible otherwise unless we did something tricky with elastic

denis-yuen reviewed (approved):
On passing build

kathy-t commented:
Updated the PR so that if there is no manual or automatic topic, the topic selection is automatically set to AI.

Furthermore, I modified the updateWorkflow and updateContainer endpoint so that it does not update the topicAI field since AI topics are updated by the topic generator, not by end users.

svonworl commented:
> Updated the PR so that if there is no manual or automatic topic, the topic selection is automatically set to AI.

Does this mean that in those cases, the AI topic automatically appears in search and/or on the entry page, without entry owner approval?

kathy-t commented:
> > Updated the PR so that if there is no manual or automatic topic, the topic selection is automatically set to AI.
> 
> Does this mean that in those cases, the AI topic automatically appears in search and/or on the entry page, without entry owner approval?

Yeah, that's correct, but it should only be for published entries. Are you thinking it should be opt-in instead of opt-out?

svonworl commented:
> > > Updated the PR so that if there is no manual or automatic topic, the topic selection is automatically set to AI.
> > 
> > 
> > Does this mean that in those cases, the AI topic automatically appears in search and/or on the entry page, without entry owner approval?
> 
> Yeah, that's correct, but it should only be for published entries. Are you thinking it should be opt-in instead of opt-out?

Mixed feelings, automatic inclusion is good for usability, but I feel like there's the possibility that owners might not always approve.  Acceptable risk?  Not sure.

denis-yuen commented:
> Mixed feelings, automatic inclusion is good for usability, but I feel like there's the possibility that owners might not always approve. Acceptable risk? Not sure.

Well, this was the intent, to have the topic sentences show up as a last resort but be labelled as AI 
https://github.com/dockstore/dockstore-ui2/pull/1939

Many workflows should have a topic one way or another, harvested from the github repo, from the .dockstore.yml, or from the workflow descriptor itself

kathy-t commented on dockstore-webservice/src/main/java/io/dockstore/webservice/api/UpdateAITopicRequest.java:
```diff
@@ -0,0 +1,42 @@
+public class UpdateAITopicRequest {
+
+    @NotNull
+    @Schema(description = "The AI topic", requiredMode = RequiredMode.REQUIRED)
```
I went with the silent truncation to match the behaviour of `topicAutomatic`, which silently truncates topics greater than `TOPIC_LENGTH`. I'll add a description saying that topics > `TOPIC_LENGTH` are truncated.

> Also, WorkflowResource.updateWorkflow(), which takes a workflow as a parameter, will I believe enforce a TOPIC_LIMIT by not deserializing a workflow with a greater than TOPIC_LENGTH ai topic.

That's true if `maxLength` is set, so perhaps it's another reason why we shouldn't set it since the AI topic can't be updated with `WorkflowResource.updateWorkflow()` anymore
